crashtracker/src/collector/crash_handler.rs

```diff
@@ -2,35 +2,243 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #![cfg(unix)]
+#![allow(deprecated)]
 
+use super::atomguard::AtomGuard;
 use super::emitters::emit_crashreport;
+use super::saguard::SaGuard;
 use crate::crash_info::CrashtrackerMetadata;
 use crate::shared::configuration::{CrashtrackerConfiguration, CrashtrackerReceiverConfig};
 use anyhow::Context;
 use libc::{
-    c_void, mmap, sigaltstack, siginfo_t, MAP_ANON, MAP_FAILED, MAP_PRIVATE, PROT_NONE, PROT_READ,
-    PROT_WRITE, SIGSTKSZ,
+    c_void, execve, mmap, sigaltstack, siginfo_t, MAP_ANON, MAP_FAILED, MAP_PRIVATE, PROT_NONE,
+    PROT_READ, PROT_WRITE, SIGSTKSZ,
 };
-use nix::sys::signal;
-use nix::sys::signal::{SaFlags, SigAction, SigHandler};
-use std::fs::File;
-use std::io::Read;
+use nix::poll::{poll, PollFd, PollFlags};
+use nix::sys::signal::{self, SaFlags, SigAction, SigHandler};
+use nix::sys::socket;
+use nix::sys::wait::{waitpid, WaitPidFlag, WaitStatus};
+use nix::unistd::{close, Pid};
+use std::ffi::CString;
+use std::fs::{File, OpenOptions};
 use std::io::Write;
-use std::os::unix::net::UnixStream;
-use std::process::{Command, Stdio};
+use std::os::unix::{
+    io::{BorrowedFd, FromRawFd, IntoRawFd, RawFd},
+    net::UnixStream,
+};
 use std::ptr;
 use std::sync::atomic::Ordering::SeqCst;
 use std::sync::atomic::{AtomicBool, AtomicPtr, AtomicU64};
+use std::time::{Duration, Instant};
+
+// Note that this file makes use the following async-signal safe functions in a signal handler.
+// <https://man7.org/linux/man-pages/man7/signal-safety.7.html>
+// - clock_gettime
+// - close (although Rust may call `free` because we call the higher-level nix interface)
+// - dup2
+// - fork (but specifically only because it does so without calling atfork handlers)
+// - kill
+// - poll
+// - raise
+// - read
+// - sigaction
+// - write
+
+// The use of fork or vfork is influenced by the availability of the function in the host libc.
+// Macos seems to have deprecated vfork.  The reason to prefer vfork is to suppress atfork
+// handlers.  This is OK because macos is primarily a test platform, and we have system-level
+// testing on Linux in various CI environments.
+#[cfg(target_os = "macos")]
+use libc::fork as vfork;
+
+#[cfg(target_os = "linux")]
+use libc::vfork;
 
 #[derive(Debug)]
 struct OldHandlers {
     pub sigbus: SigAction,
     pub sigsegv: SigAction,
 }
 
-enum ReceiverType {
-    ForkedProcess(std::process::Child),
-    UnixSocket(String),
+struct Receiver {
+    receiver_uds: RawFd,
+    receiver_pid: i32,
+}
+
+// The args_cstrings and env_vars_strings fields are just storage.  Even though they're
+// unreferenced, they're a necessary part of the struct.
+#[allow(dead_code)]
+struct PreparedExecve {
+    binary_path: CString,
+    args_cstrings: Vec<CString>,
+    args_ptrs: Vec<*const i8>,
+    env_vars_cstrings: Vec<CString>,
+    env_vars_ptrs: Vec<*const i8>,
```

Comment:
Shouldn't these Vecs be `*const libc::c_char`? Different architectures have different signedness for char:
https://github.com/rust-lang/libc/blob/81f0cd3d9715e579c92063f78d0d14da85a31dd1/src/unix/linux_like/linux/gnu/b64/aarch64/mod.rs#L3
https://github.com/rust-lang/libc/blob/81f0cd3d9715e579c92063f78d0d14da85a31dd1/src/unix/linux_like/linux/musl/b64/x86_64/mod.rs#L1

Reply:
Ah, yes.  I'm surprised I didn't get an error here, since the aliases are clearly different, but this is definitely something that's broken things before.  Good catch--thank you.

Reply:
Fixed.  I suspect this would have broken ARM builds entirely due to how the conversions work, which makes me think I ought to build on an arm-linux machine before merging in case there are other issues.